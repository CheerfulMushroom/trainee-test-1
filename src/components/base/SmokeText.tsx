import * as React from "react";

import * as generateClassName from "smokescreen/Cid";

import {
	generateStyleString,
	obfuscateString,
	generateRandomString,
} from "./utils";
import { StyleSheet } from "./utils";

export default ({ children }) => {
	const text: string = children.normalize("NFC");

	// Предупреждение 'item was found 2' (/м.*о.*с.*к.*в.*а/i.test(element.innerHTML))
	// будет возникать всегда, но не из-за SmokeText а из-за обычного текста. Например:
	// "мин. от м. Сходненская. Развитая инфраструктура. Индивидуа"

	// 1) Обфусцировать исходный текст, заменив кириллицу на английские омоглифы
	// 2) Разбить текст на части
	// 3) Каждую часть добавить в div и div::after
	// 4) Между частями контента добавить div с рандомными строками

	// Дополнительные способы максировки:
	// 1) Выводить текст в обратном порядке и в обратном направлении (unicode-bidi: bidi-override; direction: rtl;)
	// 2) Использовать вместо некоторых символов зеркальные омоглифы (напр. "И" и "N") и отзеркаливать их с помощью transform: scale(-1, 1)

	const splitText = (text: string): string[] => {
		const chars = Array.from(text);
		let chunks: string[] = [];
		const maxChunkSize = Math.floor(chars.length ** 0.5);

		let i = 0;
		while (i < chars.length) {
			const chunkSize = Math.ceil(Math.random() * maxChunkSize);
			const chunk = chars.slice(i, i + chunkSize).join("");
			chunks.push(chunk);
			i += chunkSize;
		}

		return chunks;
	};

	const obfuscatedText = obfuscateString(text);
	const sheet: StyleSheet = {};
	const elements = [];

	const textChunks = splitText(obfuscatedText);
	for (let i = 0; i < textChunks.length; ) {
		// Content div with ::after
		const className = generateClassName();
		sheet[`.${className}`] = {
			display: "inline",
		};

		elements.push(
			<div key={className} className={className}>
				{obfuscateString(textChunks[i++])}
			</div>
		);

		if (i < textChunks.length) {
			const chunk = textChunks[i++];
			sheet[`.${className}::after`] = {
				content: `"${obfuscateString(chunk)}"`,
			};
		}

		// Decoy div
		const decoyClassName = generateClassName();
		const maxDecoyLength = 3;
		const decoyContent = generateRandomString(maxDecoyLength);

		sheet[`.${decoyClassName}`] = {
			display: "none",
		};

		elements.push(
			<div key={decoyClassName} className={decoyClassName}>
				{decoyContent}
			</div>
		);
	}

	const wrapperClassName = generateClassName();
	sheet[`.${wrapperClassName}`] = {
		display: "inline",
	};

	return (
		<div className={wrapperClassName}>
			<style>{generateStyleString(sheet)}</style>
			{elements}
		</div>
	);
};
