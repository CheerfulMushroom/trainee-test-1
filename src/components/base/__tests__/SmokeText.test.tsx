import React = require("react");
import { describe, test, expect } from "@jest/globals";
const renderer = require("react-test-renderer");

import { StableRandomizer } from "./__utils__/StableRandomizer";

import SmokeText from "../SmokeText";

let counter = 0;
jest.mock("smokescreen/Cid", () => {
	return () => {
		++counter;
		return "className" + counter.toString();
	};
});

describe("Рендер", () => {
	beforeEach(() => {
		counter = 0;
		const randomizer = new StableRandomizer(1);
		jest.spyOn(global.Math, "random").mockImplementation(() =>
			randomizer.random()
		);
	});

	afterEach(() => {
		jest.spyOn(global.Math, "random").mockRestore();
	});

	test('Текст "Здесь могла быть ваша реклама"', () => {
		const component = renderer.create(
			<SmokeText>{"Здесь могла быть ваша реклама"}</SmokeText>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("Пустая строка", () => {
		const component = renderer.create(<SmokeText>{""}</SmokeText>);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("Один символ", () => {
		const component = renderer.create(<SmokeText>{"А"}</SmokeText>);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("Компоненты с одинаковым текстом рендерятся по-разному", () => {
		const adText = "Жилье в центре города";
		const component1 = renderer.create(<SmokeText>{adText}</SmokeText>);
		const component2 = renderer.create(<SmokeText>{adText}</SmokeText>);

		const snap1 = component1.toJSON();
		const snap2 = component2.toJSON();

		expect(snap1).not.toEqual(snap2);
	});
});
