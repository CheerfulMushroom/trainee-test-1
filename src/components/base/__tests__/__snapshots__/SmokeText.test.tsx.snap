// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Рендер Один символ 1`] = `
<div
  className="className3"
>
  <style>
    .className1 {display: inline}.className2 {display: none}.className3 {display: inline}
  </style>
  <div
    className="className1"
  >
    А
  </div>
  <div
    className="className2"
  >
    Вн
  </div>
</div>
`;

exports[`Рендер Пустая строка 1`] = `
<div
  className="className1"
>
  <style>
    .className1 {display: inline}
  </style>
</div>
`;

exports[`Рендер Текст "Здесь могла быть ваша реклама" 1`] = `
<div
  className="className11"
>
  <style>
    .className1 {display: inline}.className1::after {content: "ь мoг"}.className2 {display: none}.className3 {display: inline}.className3::after {content: "ыт"}.className4 {display: none}.className5 {display: inline}.className5::after {content: " вaш"}.className6 {display: none}.className7 {display: inline}.className7::after {content: "peкл"}.className8 {display: none}.className9 {display: inline}.className10 {display: none}.className11 {display: inline}
  </style>
  <div
    className="className1"
  >
    Здec
  </div>
  <div
    className="className2"
  >
    рыР
  </div>
  <div
    className="className3"
  >
    ла б
  </div>
  <div
    className="className4"
  >
    ЪгЩ
  </div>
  <div
    className="className5"
  >
    ь
  </div>
  <div
    className="className6"
  >
    е
  </div>
  <div
    className="className7"
  >
    a 
  </div>
  <div
    className="className8"
  >
    эЬ
  </div>
  <div
    className="className9"
  >
    амa
  </div>
  <div
    className="className10"
  >
    збЛ
  </div>
</div>
`;
